### PR TITLE
Cranes

### PR DESCRIPTION
**Pull request by Nick007J**

### CODE REVIEW

File: src/control/Cranes.cpp [hunk 1]
```diff
@@ -2,13 +2,670 @@
...
+	if (NumCranes >= NUM_CRANES)
+		return;
+	CCrane* pCrane = &aCranes[NumCranes];
+	pCrane->Init();
+	pCrane->m_pCraneEntity = (CBuilding*)pEntity;
+	pCrane->m_nCraneStatus = CCrane::NONE;
+	pCrane->m_fHookAngle = NumCranes; // lol wtf
+	while (pCrane->m_fHookAngle > TWOPI)
+		pCrane->m_fHookAngle -= TWOPI;
+	pCrane->m_fHookOffset = 20.0f;
```
Comment (ShFil119) [inline src/control/Cranes.cpp:75]: multiple uses, could be named with define

File: src/control/Cranes.cpp [hunk 2]
```diff
@@ -2,13 +2,670 @@
...
+		pCrane->SetHookMatrix();
+	}
+	else
+		pCrane->m_pHook = nil;
+	NumCranes++;
+}
+
+void CCranes::ActivateCrane(float fInfX, float fSupX, float fInfY, float fSupY, float fDropOffX, float fDropOffY, float fDropOffZ, float fHeading, bool bIsCrusher, bool bIsMilitary, float fPosX, float fPosY)
+{
+	float fMinDistance = 99999.9f;
```
Comment (ShFil119) [inline src/control/Cranes.cpp:102]: also

File: src/control/Cranes.cpp [hunk 3]
```diff
@@ -2,13 +2,670 @@
...
+	float X = fPosX, Y = fPosY;
+	if (X <= -10000.0f || Y <= -10000.0f) {
```
Comment (ShFil119) [inline src/control/Cranes.cpp:104]: also

File: src/control/Cranes.cpp [hunk 4]
```diff
@@ -2,13 +2,670 @@
...
+						}
+					}
+				}
+			}
+			else
+				m_nCraneState = IDLE;
+			break;
+		case LIFTING_TARGET:
+			RotateCarriedCarProperly();
+			if (GoTowardsTarget(m_fDropoffAngle, m_fDropoffDistance, GetHeightToDropoff(), 0.3f))
```
Comment (ShFil119) [inline src/control/Cranes.cpp:272]: for this you have even a name.

File: src/control/Cranes.cpp [hunk 5]
```diff
@@ -2,13 +2,670 @@
...
+			Abs(pVehicle->GetMoveSpeed().z) >= CAR_MOVING_SPEED_THRESHOLD)
+			continue;
+		if (!pVehicle->IsCar() || pVehicle->m_status == STATUS_WRECKED || pVehicle->m_fHealth < 250.0f)
+			continue;
+		if (!DoesCranePickUpThisCarType(pVehicle->GetModelIndex()) ||
+			m_bIsMilitaryCrane && CCranes::DoesMilitaryCraneHaveThisOneAlready(pVehicle->GetModelIndex())) {
+			if (!pVehicle->bCraneMessageDone) {
+				pVehicle->bCraneMessageDone = true;
+				if (!m_bIsMilitaryCrane)
+					CGarages::TriggerMessage("CR_1", -1, 4000, -1); // Crane cannot lift this vehicle.
```
Comment (ShFil119) [inline src/control/Cranes.cpp:418]: also (4000)

**Comment by aap** [comment]
very nice

**Comment by erorcun** [comment]
lgbt

**Pull request merged**